examples/plot_credit_loan_decisions.py

```diff
@@ -0,0 +1,1268 @@
+# However, this does come with the trade-off that the
+# :code:`ThresholdOptimizer` achieves a lower :code:`balanced_accuracy` score
+# for *male* applicants.
+#
+# Reductions approach to unfairness mitigation
+# --------------------------------------------
+# In the previous section, we took a fairness-unaware model and used the
+# :code:`ThresholdOptimizer` to transform the model's decision boundary to
+# satisfy our fairness constraints.
+# One key limitation of the :code:`ThresholdOptimizer` is needing access to
+# our *sensitive_feature* during prediction time.
+#
+# In this section, we will use the *reductions* approach of `Agarwal et. al
+# (2018) <https://arxiv.org/pdf/1803.02453.pdf>`_ to produce models that satisfy
+# the fairness constraint without needing access to the sensitive features at
+# deployment time.
+#
+# The main reduction algorithm in Fairlearn is :code:`ExponentiatedGradient`.
+# The algorithm creates a sequence of re-weighted datasets and retrains the
+# wrapped classifier on each of the datasets.
+# This re-training process is guaranteed to find a model that satisfies the
+# fairness constraints while optimizing the performance metric.
+#
+# The model returned by :code:`ExponentiatedGradient` consists of several inner
+# models, returned by a wrapped estimator.
+#
+# To instantiate an :code:`ExponentiatedGradient` model, we pass in two
+# parameters:
+#
+# - a base :code:`estimator` (object that supports training)
+# - fairness :code:`constraints` (object of type :class:`fairlearn.reductions.Moment`)
+#
+# When passing in a fairness *constraint* as a :code:`Moment`, we can specify
+# an :code:`epsilon` value representing the maximum allowed difference or ratio
+# between our largest and smallest value.
+# For example, in the below code,
+# :code:`EqualizedOdds(difference_bound=epsilon)` means that we are using
+# :code:`EqualizedOdds` as our fairness constraint, and we will allow a
+# maximal difference of :code:`epsilon` between our largest and smallest
+# *equalized odds* value.
+
+def get_expgrad_models_per_epsilon(
+    estimator, epsilon, X_train, y_train, A_train
+):
+    """Instantiate and train an ExponentiatedGradient model on the
+    balanced training dataset.
+
+    Parameters
+    ----------
+    Estimator: Base estimator to contains a fit and predict function.
+    Epsilon: Float representing maximum difference bound for the fairness Moment constraint
+
+    Returns
+    -------
+    Predictors
+        List of inner model predictors learned by the ExponentiatedGradient
+        model during the training process.
+
+    """
+    exp_grad_est = ExponentiatedGradient(
+        estimator=estimator.named_steps["classifier"],
+        constraints=EqualizedOdds(difference_bound=epsilon),
+    )
+    # Is this an issue - Re-runs
+    exp_grad_est.fit(X_train, y_train, sensitive_features=A_train)
+    predictors = exp_grad_est.predictors_
+    return predictors
```

Comment:
I did try to look into the CircleCI issue. It runs properly till we encounter the function `get_expgrad_models_per_epsilon` call.

The iterations get processed very slowly, not sure why it happens so quickly on local. Maybe we could limit it somehow, or make this part hard-coded at least for generating sphinx gallery examples.

Reply:
There's been a little bit of a discussion about this on [Discord](https://discordapp.com/channels/840099830160031744/872866935610703952/1060658509861109930), too. I'm still wondering if we should stick with CircleCI at all, assuming this issue wouldn't happen on GitHub Actions. But migrating is also work, of course.

Reply:
I've been fiddling with this on and off, but still no luck.

Reply:
Now is now resolved. See:
https://github.com/fairlearn/fairlearn/pull/1170#issuecomment-1377421962